### PR TITLE
Prefer `expected == atomic_compare_exchange(ptr, expected, desired)`

### PR DESCRIPTION
**Pull request by dalg24**
Rewrote `atomic_compare_exchange_strong(ptr, expected, desired)` as `expected == atomic_compare_exchange(ptr, expected, desired)` as `atomic_compare_exchange_strong` is on its way to be deprecated/removed (see https://github.com/kokkos/kokkos/pull/7458)

### CODE REVIEW

**Review by lucbv — APPROVED** [review]

**Review by cwpearson — CHANGES_REQUESTED** [review]
If I understand correctly, `atomic_compare_exchange_strong` updates the `expected` value (it takes it by reference), whereas `atomic_compare_exchange` does not, so these re-writes are not equivalent.

**Comment by dalg24** [comment]
> If I understand correctly, `atomic_compare_exchange_strong` updates the `expected` value (it takes it by reference), whereas `atomic_compare_exchange` does not, so these re-writes are not equivalent.

No, it takes it by value
https://github.com/kokkos/kokkos/blob/3fe17d96d42eb2bd84f1f2a835a2ed0ba87606b6/core/src/Kokkos_Atomics_Desul_Wrapper.hpp#L135-L139
(I know it is confusing.  The desul counterpart does but I will propose to overhaul this later.  It will be transparent to Kokkos.)
The `atomic_compare_exchange` semantics are easier to reason about.  They follow atomicCAS() form CUDA.

**Comment by dalg24** [comment]
Here is a link to the latest release (before I started messing with the API) to convince yourself that this is right.
It is unchanged since 4.0

**Review by cwpearson — APPROVED** [review]
I see now, thanks. I'm good with it once the tests are passing! Thanks @dalg24

**Comment by ndellingwood** [comment]
Lots of failing kokkos-kernels nightlies with e.g.
```
05:34:15 /home/jenkins/blake-new/workspace/KokkosKernels_Nightly_Blake_OneAPI_2023_1_0_Sycl_PV-oneMKL/kokkos-kernels/common/src/KokkosKernels_HashmapAccumulator.hpp:519:23: error: no member named 'atomic_compare_exchange_strong' in namespace 'Kokkos'
05:34:15           if (Kokkos::atomic_compare_exchange_strong(keys + hash, -1, key)) {
```

I'm assuming this PR will resolve them once ready for merge. Cross-referencing https://github.com/kokkos/kokkos/pull/7458

**Comment by ndellingwood** [comment]
Multiple jobs failing with timeout in the following tests:
```
 	 14 - graph_cuda (Timeout)
	 16 - sparse_cuda (Timeout)
```

**graph_cuda** seems to be hanging in the test
```
 [ RUN      ] Cuda.graph_random_graph_coarsen_double_int_int_TestDevice
```

**sparse_cuda** seems to be hanging in the test
```
[ RUN      ] Cuda.sparse_spgemm_double_int_int_TestDevice
```

File: sparse/impl/KokkosSparse_spgemm_impl_kkmem.hpp [hunk 1]
```diff
@@ -687,12 +687,11 @@ struct KokkosSPGEMM<HandleType, a_row_view_t_, a_lno_nnz_view_t_, a_scalar_nnz_v
             if (keys[trial] == my_b_col) {
               Kokkos::atomic_add(vals + trial, my_b_val);
               fail = 0;
-              break;
```
Comment (dalg24) [inline sparse/impl/KokkosSparse_spgemm_impl_kkmem.hpp:690]: ugggg

File: sparse/impl/KokkosSparse_spgemm_impl_kkmem.hpp [hunk 2]
```diff
@@ -687,12 +687,11 @@ struct KokkosSPGEMM<HandleType, a_row_view_t_, a_lno_nnz_view_t_, a_scalar_nnz_v
             if (keys[trial] == my_b_col) {
               Kokkos::atomic_add(vals + trial, my_b_val);
               fail = 0;
```
Comment (dalg24) [inline sparse/impl/KokkosSparse_spgemm_impl_kkmem.hpp:689]:
```suggestion
              fail = 0;
              break;
```

**Review by ndellingwood — APPROVED** [review]
Tentative approval to trigger AT2

**Comment by ndellingwood** [comment]
@dalg24 can you add the dco signoff as well (check failed)? Sorry for the inconvenience

**Pull request merged**